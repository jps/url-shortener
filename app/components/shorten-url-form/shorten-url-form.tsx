import { yupResolver } from "@hookform/resolvers/yup";
import { useState } from "react";
import { useForm } from "react-hook-form";
import * as yup from "yup";

export interface ShortenFormData {
  url: string;
}

export interface ShortenUrlFormProps {
  onSubmit: (data: ShortenFormData) => void;
}

/*
Note - there is some potential complexity here... the default yup rule will do a good enough job but the UX could be 
improved as lot of users will potentially enter a protocoless url like: something.com in this scenario we could 
either feedback to the user as we do currently or a default protocol could be added.
*/
const schema = yup
  .object({
    url: yup.string().url("Please enter a valid url").max(4096).required(),
  })
  .required();

export const ShortenUrlForm = ({ onSubmit }: ShortenUrlFormProps) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
    reset,
  } = useForm<ShortenFormData>({
    resolver: yupResolver(schema),
  });

  const [isSubmitting, setIsSubmitting] = useState(false);

  const onSubmitHandler = async (data: ShortenFormData) => {
    try {
      setIsSubmitting(true);
      onSubmit(data);
      reset();
    } catch (exception) {
      //TODO: feedback to user if there is a server error
    }
    setIsSubmitting(false);
  };

  return (
    <form className="shorten-form" onSubmit={handleSubmit(onSubmitHandler)}>
      <h2 className="shorten-form__title">Shorten URL</h2>
      <label className="visually-hidden" htmlFor="url">
        Url
      </label>
      {errors.url && (
        <div
          role="alert"
          className="shorten-form__validation form-validation-message"
        >
          {errors.url.message}
        </div>
      )}
      <input
        id="url"
        className={`shorten-form__input form-text-input${
          errors.url ? " form-text-input--invalid" : ""
        }`}
        {...register("url")}
        placeholder="Shorten your link e.g. https://example.com"
        required
      />
      <input
        className="shorten-form__save form-submit"
        type="submit"
        value="Shorten"
        disabled={isSubmitting}
      />
    </form>
  );
};
