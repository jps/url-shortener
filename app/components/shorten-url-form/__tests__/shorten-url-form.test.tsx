import { render, screen, waitFor } from "@testing-library/react";
import { ShortenUrlForm } from "../shorten-url-form";
import userEvent from "@testing-library/user-event";

describe("ShortenUrlForm should", () => {
  it("render as expected", () => {
    const { asFragment } = render(<ShortenUrlForm onSubmit={jest.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("submit form for valid url", async () => {
    const onSubmit = jest.fn();
    userEvent.setup();

    render(<ShortenUrlForm onSubmit={onSubmit} />);

    const textInput = screen.getByLabelText<HTMLInputElement>("Url");
    const submitButton = screen.getByText("Shorten");
    const validUrl = "https://example.com";

    await userEvent.type(textInput, validUrl);
    await userEvent.click(submitButton);

    expect(onSubmit).toHaveBeenCalledTimes(1);
    expect(onSubmit).toHaveBeenCalledWith({ url: validUrl });
    expect(textInput.value).toBe("");
  });

  it("show error for invalid url", async () => {
    const onSubmit = jest.fn();
    userEvent.setup();

    render(<ShortenUrlForm onSubmit={onSubmit} />);

    const textInput = screen.getByLabelText("Url");
    const submitButton = screen.getByText("Shorten");

    await userEvent.type(textInput, "0123456789");
    await userEvent.click(submitButton);

    await waitFor(() => screen.getByText("Please enter a valid url"));

    expect(onSubmit).toHaveBeenCalledTimes(0);
  });
});
