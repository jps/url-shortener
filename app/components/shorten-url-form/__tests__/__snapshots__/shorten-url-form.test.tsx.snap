// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShortenUrlForm should render as expected 1`] = `
<DocumentFragment>
  <form
    class="shorten-form"
  >
    <h2
      class="shorten-form__title"
    >
      Shorten URL
    </h2>
    <label
      class="visually-hidden"
      for="url"
    >
      Url
    </label>
    <input
      class="shorten-form__input form-text-input"
      id="url"
      name="url"
      placeholder="Shorten your link e.g. https://example.com"
      required=""
    />
    <input
      class="shorten-form__save form-submit"
      type="submit"
      value="Shorten"
    />
  </form>
</DocumentFragment>
`;
